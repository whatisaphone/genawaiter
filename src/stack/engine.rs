use crate::{core, core::Next};
use ::core::{cell::UnsafeCell, ptr};

/// This type holds the value that is pending being returned from the generator.
///
/// # Safety
///
/// This type is `!Sync` (so, single-thread), never exposed to user-land code,
/// and never borrowed across a function call, so safety can be verified locally
/// at each use site.
pub struct Airlock<Y, R>(UnsafeCell<Next<Y, R>>);

impl<Y, R> Default for Airlock<Y, R> {
    fn default() -> Self {
        Self(UnsafeCell::new(Next::Empty))
    }
}

impl<'s, Y, R> core::Airlock for &'s Airlock<Y, R> {
    type Yield = Y;
    type Resume = R;

    fn peek(&self) -> Next<(), ()> {
        // Safety: This follows the safety rules above.
        let inner = unsafe { &*self.0.get() };
        inner.without_values()
    }

    fn replace(
        &self,
        next: Next<Self::Yield, Self::Resume>,
    ) -> Next<Self::Yield, Self::Resume> {
        // Safety: This follows the safety rules above.
        unsafe { ptr::replace(self.0.get(), next) }
    }
}

/// This object lets you yield values from the generator by calling the `yield_`
/// method.
///
/// "Co" can stand for either _controller_ or _coroutine_, depending on how
/// theoretical you are feeling.
///
/// [_See the module-level docs for examples._](.)
pub type Co<'y, Y, R = ()> = core::Co<&'y Airlock<Y, R>>;
