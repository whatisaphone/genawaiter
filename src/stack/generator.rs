use core::{future::Future, mem::MaybeUninit, pin::Pin, ptr};

use crate::{
    core::{advance, async_advance, Airlock as _, Next},
    ops::{Generator, GeneratorState},
    stack::engine::{Airlock, Co},
};

/// This data structure holds the transient state of an executing generator.
///
/// It's called "Shelf", rather than "State", to avoid confusion with the
/// `GeneratorState` enum.
///
/// [_See the module-level docs for examples._](.)
pub struct Shelf<Y, R, F: Future> {
    airlock: Airlock<Y, R>,
    future: MaybeUninit<F>,
}

impl<Y, R, F: Future> Shelf<Y, R, F> {
    /// Creates a new, empty `Shelf`.
    ///
    /// [_See the module-level docs for examples._](.)
    #[must_use]
    pub fn new() -> Self {
        Self {
            airlock: Airlock::default(),
            // Safety: The lifetime of the data is controlled by a `Gen`, which constructs
            // it in place, and holds a mutable reference right up until dropping it in
            // place. Thus, the data inside is pinned and can never be moved.
            future: MaybeUninit::uninit(),
        }
    }
}

impl<Y, R, F: Future> Default for Shelf<Y, R, F> {
    #[must_use]
    fn default() -> Self {
        Self::new()
    }
}

/// This is a generator which can be stack-allocated.
///
/// [_See the module-level docs for examples._](.)
pub struct Gen<'s, Y, R, F: Future> {
    airlock: &'s Airlock<Y, R>,
    future: Pin<&'s mut F>,
}

impl<'s, Y, R, F: Future> Gen<'s, Y, R, F> {
    /// Creates a new generator from a function.
    ///
    /// The state of the generator is stored in `shelf`, which will be pinned in
    /// place while this generator exists. The generator itself is movable,
    /// since it just holds a reference to the pinned state.
    ///
    /// The function accepts a [`Co`] object, and returns a future. Every time
    /// the generator is resumed, the future is polled. Each time the future is
    /// polled, it should do one of two things:
    ///
    /// - Call `co.yield_()`, and then return `Poll::Pending`.
    /// - Drop the `Co`, and then return `Poll::Ready`.
    ///
    /// Typically this exchange will happen in the context of an `async fn`.
    ///
    /// # Safety
    ///
    /// The `Co` object must not outlive the returned `Gen`. By time the
    /// generator completes (i.e., by time the producer's Future returns
    /// `Poll::Ready`), the `Co` object should already have been dropped. If
    /// this invariant is not upheld, memory unsafety can result.
    ///
    /// Afaik, the Rust compiler [is not flexible enough][hrtb-thread] to let
    /// you express this invariant in the type system, but I would love to be
    /// proven wrong!
    ///
    /// [hrtb-thread]: https://users.rust-lang.org/t/hrtb-on-multiple-generics/34255
    ///
    /// # Examples
    ///
    /// ```rust
    /// # use genawaiter::stack::{Co, Gen, Shelf};
    /// #
    /// # async fn producer(co: Co<'_, i32>) { /* ... */ }
    /// #
    /// let mut shelf = Shelf::new();
    /// let gen = unsafe { Gen::new(&mut shelf, producer) };
    /// ```
    pub unsafe fn new(
        shelf: &'s mut Shelf<Y, R, F>,
        producer: impl FnOnce(Co<'s, Y, R>) -> F,
    ) -> Self {
        // By splitting the mutable `shelf` into a shared `airlock` and a unique
        // pinned `future` reference we ensure the aliasing rules are not violated.
        let airlock = &shelf.airlock;
        // Safety: Initializes the future in-place using `ptr::write`, which is
        // the correct way to initialize a `MaybeUninit`
        shelf.future.as_mut_ptr().write(producer(Co::new(airlock)));
        // Safety: The `MaybeUninit` is initialized by now, so its safe to create
        // a reference to the future itself
        // NB: can be replaced by `MaybeUninit::get_mut` once stabilized
        let init = &mut *shelf.future.as_mut_ptr();
        // Safety: The `shelf` remains borrowed during the entire lifetime of
        // the `Gen`and is hence pinned.
        Self {
            airlock,
            future: Pin::new_unchecked(init),
        }
    }

    /// Resumes execution of the generator.
    ///
    /// `arg` is the resume argument. If the generator was previously paused by
    /// awaiting a future returned from `co.yield()`, that future will complete,
    /// and return `arg`.
    ///
    /// If the generator yields a value, `Yielded` is returned. Otherwise,
    /// `Completed` is returned.
    ///
    /// [_See the module-level docs for examples._](.)
    pub fn resume(&mut self, arg: R) -> GeneratorState<Y, F::Output> {
        self.airlock.replace(Next::Resume(arg));
        advance(self.future.as_mut(), &self.airlock)
    }

    /// Resumes execution of the generator.
    ///
    /// If the generator pauses without yielding, `Poll::Pending` is returned.
    /// If the generator yields a value, `Poll::Ready(Yielded)` is returned.
    /// Otherwise, `Poll::Ready(Completed)` is returned.
    ///
    /// [_See the module-level docs for examples._](.)
    pub fn async_resume(
        &mut self,
        arg: R
    ) -> impl Future<Output = GeneratorState<Y, F::Output>> + '_ {
        self.airlock.replace(Next::Resume(arg));
        async_advance(self.future.as_mut(), self.airlock)
    }
}

impl<'s, Y, R, F: Future> Drop for Gen<'s, Y, R, F> {
    fn drop(&mut self) {
        // Safety: `future` itself is a `MaybeUninit`, which is guaranteed to be
        // initialized, because the only way to construct a `Gen` is with
        // `Gen::new`, which initializes it.
        //
        // Drop `future` in place first (it likely contains a reference to airlock).
        // Since we drop it in place, the `Pin` invariants are not violated.
        // The airlock is regularly dropped when the `Shelf` goes out of scope.
        unsafe {
            ptr::drop_in_place(self.future.as_mut().get_unchecked_mut());
        }
    }
}

impl<'s, Y, R, F: Future> Generator<R> for Gen<'s, Y, R, F> {
    type Yield = Y;
    type Return = F::Output;

    fn resume(
        self: Pin<&mut Self>,
        arg: R,
    ) -> GeneratorState<Self::Yield, Self::Return> {
        // Safety: `Gen::resume` does not move `self`.
        let this = unsafe { self.get_unchecked_mut() };
        this.resume(arg)
    }
}
