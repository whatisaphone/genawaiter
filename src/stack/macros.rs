/**
Creates a generator on the stack.

This macro is deprecated. Use [`let_gen!`] or [`let_gen_using!`] instead.

[`let_gen!`]: stack/macro.let_gen.html
[`let_gen_using!`]: stack/macro.let_gen_using.html
*/
#[macro_export]
#[deprecated = "Use `let_gen_using!()` instead."]
macro_rules! generator_mut {
    ($name:ident, $producer:expr $(,)?) => {
        $crate::stack::let_gen_using!($name, $producer);
    };
}

/**
Creates a generator on the stack unsafely.

This macro is deprecated. Use [`let_gen!`] or [`let_gen_using!`] instead.

[`let_gen!`]: stack/macro.let_gen.html
[`let_gen_using!`]: stack/macro.let_gen_using.html
*/
#[macro_export]
#[deprecated = "Use `let_gen_using!()` instead."]
macro_rules! unsafe_create_generator {
    ($name:ident, $producer:expr $(,)?) => {
        let mut generator_state = $crate::stack::Shelf::new();
        #[allow(unused_mut)]
        let mut $name =
            unsafe { $crate::stack::Gen::new(&mut generator_state, $producer) };
    };
}

#[cfg(test)]
mod tests {
    use crate::{
        ops::GeneratorState,
        stack::{Co, Gen, Shelf},
    };

    /// This test proves that `Gen::new` is actually unsafe.
    #[test]
    #[ignore = "compile-only test"]
    fn unsafety() {
        async fn shenanigans(co: Co<'_, i32>) -> Co<'_, i32> {
            co
        }

        let mut shelf = Shelf::new();
        let mut gen = unsafe { Gen::new(&mut shelf, shenanigans) };

        // Get the `co` out of the generator (don't try this at home).
        let mut escaped_co = match gen.resume() {
            GeneratorState::Yielded(_) => panic!(),
            GeneratorState::Complete(co) => co,
        };
        // Drop the generator. This drops the airlock (inside the state), but `co` still
        // holds a reference to the airlock.
        drop(gen);
        // Now we're able to use an invalidated reference.
        let _ = escaped_co.yield_(10);
    }
}
