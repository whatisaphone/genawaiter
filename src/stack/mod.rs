/*!
This module implements a generator which doesn't allocate.

You can create a basic generator with [`let_gen!`] and [`yield_!`].

[`let_gen!`]: macro.let_gen.html

```rust
# #[cfg(feature = "proc_macro")]
# fn feature_gate() {
# use genawaiter::{stack::let_gen, yield_};
#
let_gen!(my_generator, {
    yield_!(10);
});
# my_generator.resume();
# }
```

If you don't like macros, you can use the low-level API directly, though note that this
requires you to trade away safety.

```rust
# use genawaiter::stack::{Co, Gen, Shelf};
#
async fn my_producer(mut co: Co<'_, u8>) {
    co.yield_(10).await;
}
let mut shelf = Shelf::new();
let mut my_generator = unsafe { Gen::new(&mut shelf, my_producer) };
# my_generator.resume();
```

# Examples

## Using `Iterator`

Generators implement `Iterator`, so you can use them in a for loop:

```rust
# #[cfg(feature = "proc_macro")]
# fn feature_gate() {
use genawaiter::{stack::let_gen, yield_};

let_gen!(odds_under_ten, {
    let mut n = 1;
    while n < 10 {
        yield_!(n);
        n += 2;
    }
});

# let mut test = Vec::new();
for num in odds_under_ten {
    println!("{}", num);
    # test.push(num);
}
# assert_eq!(test, [1, 3, 5, 7, 9]);
# }
```

## Collecting into a `Vec`

```rust
# #[cfg(feature = "proc_macro")]
# fn feature_gate() {
# use genawaiter::{stack::let_gen, yield_};
#
# let_gen!(odds_under_ten, {
#     for n in (1..).step_by(2).take_while(|&n| n < 10) { yield_!(n); }
# });
#
let xs: Vec<_> = odds_under_ten.into_iter().collect();
assert_eq!(xs, [1, 3, 5, 7, 9]);
# }
```

## A generator is a closure

Like any closure, you can capture values from outer scopes.

```rust
# #[cfg(feature = "proc_macro")]
# fn feature_gate() {
# use genawaiter::{stack::let_gen, yield_, GeneratorState};
#
let two = 2;
let_gen!(multiply, {
    yield_!(10 * two);
});
assert_eq!(multiply.resume(), GeneratorState::Yielded(20));
# }
```

## Using `resume()`

```rust
# #[cfg(feature = "proc_macro")]
# fn feature_gate() {
# use genawaiter::{stack::let_gen, yield_, GeneratorState};
#
# let_gen!(odds_under_ten, {
#     for n in (1..).step_by(2).take_while(|&n| n < 10) { yield_!(n); }
# });
#
assert_eq!(odds_under_ten.resume(), GeneratorState::Yielded(1));
assert_eq!(odds_under_ten.resume(), GeneratorState::Yielded(3));
assert_eq!(odds_under_ten.resume(), GeneratorState::Yielded(5));
assert_eq!(odds_under_ten.resume(), GeneratorState::Yielded(7));
assert_eq!(odds_under_ten.resume(), GeneratorState::Yielded(9));
assert_eq!(odds_under_ten.resume(), GeneratorState::Complete(()));
# }
```

## Passing resume arguments

You can pass values into the generator.

Note that the first resume argument will be lost. This is because at the time the first
value is sent, there is no future being awaited inside the generator, so there is no
place the value could go where the generator could observe it.

```rust
# #[cfg(feature = "proc_macro")]
# fn feature_gate() {
# use genawaiter::{stack::let_gen, yield_};
#
let_gen!(check_numbers, {
    let num = yield_!(());
    assert_eq!(num, 1);

    let num = yield_!(());
    assert_eq!(num, 2);
});

check_numbers.resume_with(0);
check_numbers.resume_with(1);
check_numbers.resume_with(2);
# }
```

## Returning a completion value

You can return a completion value with a different type than the values that are
yielded.

```rust
# #[cfg(feature = "proc_macro")]
# fn feature_gate() {
# use genawaiter::{stack::let_gen, yield_, GeneratorState};
#
let_gen!(numbers_then_string, {
    yield_!(10);
    yield_!(20);
    "done!"
});

assert_eq!(numbers_then_string.resume(), GeneratorState::Yielded(10));
assert_eq!(numbers_then_string.resume(), GeneratorState::Yielded(20));
assert_eq!(numbers_then_string.resume(), GeneratorState::Complete("done!"));
# }
```

## Defining a reusable producer function

```rust
# #[cfg(feature = "proc_macro")]
# fn feature_gate() {
# use genawaiter::{stack::{let_gen_using, producer_fn}, yield_, GeneratorState};
#
#[producer_fn(u8)]
async fn produce() {
    yield_!(10);
}

let_gen_using!(gen, produce);
assert_eq!(gen.resume(), GeneratorState::Yielded(10));
# }
```

## Using the low-level API

You can define an `async fn` directly, instead of relying on the `gen!` or `producer!`
macros.

```rust
use genawaiter::stack::{let_gen_using, Co};

async fn producer(mut co: Co<'_, i32>) {
    let mut n = 1;
    while n < 10 {
        co.yield_(n).await;
        n += 2;
    }
}

let_gen_using!(odds_under_ten, producer);
let result: Vec<_> = odds_under_ten.into_iter().collect();
assert_eq!(result, [1, 3, 5, 7, 9]);
```

## Using the low-level API with an async closure (nightly Rust only)

```ignore
# use genawaiter::{stack::let_gen_using, GeneratorState};
#
let_gen_using!(gen, async move |co| {
    co.yield_(10).await;
    co.yield_(20).await;
});
assert_eq!(gen.resume(), GeneratorState::Yielded(10));
assert_eq!(gen.resume(), GeneratorState::Yielded(20));
assert_eq!(gen.resume(), GeneratorState::Complete(()));
```

## Using the low-level API with an async <del>closure</del> fauxÂ·sure (for stable Rust)

```
# use genawaiter::{stack::let_gen_using, GeneratorState};
#
let_gen_using!(gen, |mut co| async move {
    co.yield_(10).await;
    co.yield_(20).await;
});
assert_eq!(gen.resume(), GeneratorState::Yielded(10));
assert_eq!(gen.resume(), GeneratorState::Yielded(20));
assert_eq!(gen.resume(), GeneratorState::Complete(()));
```

## Using the low-level API with function arguments

This is just ordinary Rust, nothing special.

```rust
# use genawaiter::{stack::{let_gen_using, Co}, GeneratorState};
#
async fn multiples_of(num: i32, mut co: Co<'_, i32>) {
    let mut cur = num;
    loop {
        co.yield_(cur).await;
        cur += num;
    }
}

let_gen_using!(gen, |co| multiples_of(10, co));
assert_eq!(gen.resume(), GeneratorState::Yielded(10));
assert_eq!(gen.resume(), GeneratorState::Yielded(20));
assert_eq!(gen.resume(), GeneratorState::Yielded(30));
```
*/

pub use crate::stack::{
    engine::Co,
    generator::{Gen, Shelf},
};

/// Creates a generator.
///
/// The first argument is the name of the resulting variable.
///
/// ```ignore
/// let_gen!(my_generator, { /* ... */ });
/// // Think of this as the spiritual equivalent of:
/// let mut my_generator = Gen::new(/* ... */);
/// ```
///
/// The second argument is the body of the generator. It should contain one or
/// more calls to the [`yield_!`] macro.
///
/// This macro is a shortcut for creating both a generator and its backing state
/// (called a [`Shelf`](struct.Shelf.html)). If you (or your IDE) dislike
/// macros, you can also do the bookkeeping by hand by using
/// [`Gen::new`](struct.Gen.html#method.new), though note that this requires you
/// to trade away safety.
///
/// # Examples
///
/// [_See the module-level docs for examples._](.)
#[cfg(feature = "proc_macro")]
pub use genawaiter_macro::stack_let_gen as let_gen;

/// Creates a generator using a producer defined elsewhere.
///
/// The first argument is the name of the resulting variable.
///
/// ```ignore
/// let_gen!(my_generator, { /* ... */ });
/// // Think of this as the spiritual equivalent of:
/// let mut my_generator = Gen::new(/* ... */);
/// ```
///
/// The second line is the producer that will be used. It can be one of these
/// two things:
///
/// 1.  The result of [`stack_producer!`] or [`stack_producer_fn!`]
///
///     [`stack_producer_fn!`]: attr.producer_fn.html
///
/// 2.  A function with this type:
///
///     ```ignore
///     async fn producer(co: Co<'_, Yield, Resume>) -> Completion { /* ... */ }
///     // which is equivalent to:
///     fn producer(co: Co<'_, Yield, Resume>) -> impl Future<Output = Completion> { /* ... */ }
///     ```
///
/// This macro is a shortcut for creating both a generator and its backing state
/// (called a [`Shelf`](struct.Shelf.html)). If you (or your IDE) dislike
/// macros, you can also do the bookkeeping by hand by using
/// [`Gen::new`](struct.Gen.html#method.new), though note that this requires you
/// to trade away safety.
///
/// # Examples
///
/// [_See the module-level docs for examples._](.)
pub use genawaiter_macro::stack_let_gen_using as let_gen_using;

/// Turns a function into a producer, which can then be used to create a
/// generator.
///
/// The body of the function should contain one or more [`yield_!`] expressions.
///
/// # Examples
///
/// [_See the module-level docs for examples._](.)
#[cfg(feature = "proc_macro")]
pub use genawaiter_proc_macro::stack_producer_fn as producer_fn;

#[macro_use]
mod macros;
mod engine;
mod generator;
mod iterator;
#[cfg(feature = "futures03")]
mod stream;

#[cfg(feature = "nightly")]
#[cfg(test)]
mod nightly_tests;

#[cfg(test)]
mod tests {
    use crate::{
        stack::{let_gen_using, Co},
        testing::DummyFuture,
        GeneratorState,
    };
    use std::{
        cell::RefCell,
        sync::{
            atomic::{AtomicBool, Ordering},
            Arc,
        },
    };

    async fn simple_producer(mut co: Co<'_, i32>) -> &'static str {
        co.yield_(10).await;
        "done"
    }

    #[test]
    fn function() {
        let_gen_using!(gen, simple_producer);
        assert_eq!(gen.resume(), GeneratorState::Yielded(10));
        assert_eq!(gen.resume(), GeneratorState::Complete("done"));
    }

    #[test]
    fn simple_closure() {
        async fn gen(i: i32, mut co: Co<'_, i32>) -> &'static str {
            co.yield_(i * 2).await;
            "done"
        }

        let_gen_using!(gen, |co| gen(5, co));
        assert_eq!(gen.resume(), GeneratorState::Yielded(10));
        assert_eq!(gen.resume(), GeneratorState::Complete("done"));
    }

    #[test]
    fn resume_args() {
        async fn gen(resumes: &RefCell<Vec<&str>>, mut co: Co<'_, i32, &'static str>) {
            let resume_arg = co.yield_(10).await;
            resumes.borrow_mut().push(resume_arg);
            let resume_arg = co.yield_(20).await;
            resumes.borrow_mut().push(resume_arg);
        }

        let resumes = RefCell::new(Vec::new());
        let_gen_using!(gen, |co| gen(&resumes, co));
        assert_eq!(*resumes.borrow(), &[] as &[&str]);

        assert_eq!(gen.resume_with("ignored"), GeneratorState::Yielded(10));
        assert_eq!(*resumes.borrow(), &[] as &[&str]);

        assert_eq!(gen.resume_with("abc"), GeneratorState::Yielded(20));
        assert_eq!(*resumes.borrow(), &["abc"]);

        assert_eq!(gen.resume_with("def"), GeneratorState::Complete(()));
        assert_eq!(*resumes.borrow(), &["abc", "def"]);
    }

    #[test]
    #[should_panic(expected = "non-async method")]
    fn forbidden_await_helpful_message() {
        async fn wrong(_: Co<'_, i32>) {
            DummyFuture.await;
        }

        let_gen_using!(gen, wrong);
        gen.resume();
    }

    #[test]
    #[should_panic(expected = "Co::yield_")]
    fn multiple_yield_helpful_message() {
        async fn wrong(mut co: Co<'_, i32>) {
            let _ = co.yield_(10);
            let _ = co.yield_(20);
        }

        let_gen_using!(gen, wrong);
        gen.resume();
    }

    #[test]
    #[should_panic = "should have been dropped by now"]
    fn escaped_co_helpful_message() {
        async fn shenanigans(co: Co<'_, i32>) -> Co<'_, i32> {
            co
        }

        let_gen_using!(gen, shenanigans);
        let mut escaped_co = match gen.resume() {
            GeneratorState::Yielded(_) => panic!(),
            GeneratorState::Complete(co) => co,
        };
        let _ = escaped_co.yield_(10);
    }

    /// Test the unsafe `Gen::drop` implementation.
    #[test]
    fn test_gen_drop() {
        struct SetFlagOnDrop(Arc<AtomicBool>);

        impl Drop for SetFlagOnDrop {
            fn drop(&mut self) {
                self.0.store(true, Ordering::SeqCst);
            }
        }

        let flag = Arc::new(AtomicBool::new(false));
        {
            let capture_the_flag = flag.clone();
            let_gen_using!(gen, |mut co| {
                async move {
                    let _set_on_drop = SetFlagOnDrop(capture_the_flag);
                    co.yield_(10).await;
                    // We will never make it this far.
                    unreachable!();
                }
            });
            assert_eq!(gen.resume(), GeneratorState::Yielded(10));
            // `gen` is only a reference to the generator, and dropping a reference has
            // no effect. The underlying generator is hidden behind macro hygiene and so
            // cannot be dropped early.
            #[allow(clippy::drop_ref)]
            drop(gen);
            assert_eq!(flag.load(Ordering::SeqCst), false);
        }
        // After the block above ends, the generator goes out of scope and is dropped,
        // which drops the incomplete future, which drops `_set_on_drop`, which sets the
        // flag.
        assert_eq!(flag.load(Ordering::SeqCst), true);
    }
}
