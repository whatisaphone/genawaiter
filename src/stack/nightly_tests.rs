// These tests can't be parsed on non-nightly compilers, so move them to a
// separate file.

use crate::{ops::GeneratorState, stack::let_gen_using};

#[test]
fn async_closure() {
    let_gen_using!(gen, async move |mut co| {
        co.yield_(10).await;
        "done"
    });
    assert_eq!(gen.resume(), GeneratorState::Yielded(10));
    assert_eq!(gen.resume(), GeneratorState::Complete("done"));
}
