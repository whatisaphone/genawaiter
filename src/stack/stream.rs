use crate::{ops::GeneratorState, stack::Gen};
use futures_core::{
    task::{Context, Poll},
    Stream,
};
use std::{future::Future, pin::Pin};

impl<'s, Y, F: Future<Output = ()>> Stream for Gen<'s, Y, (), F> {
    type Item = Y;

    fn poll_next(
        mut self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Option<Self::Item>> {
        let fut = self.async_resume();
        pin_mut!(fut);
        match fut.poll(cx) {
            Poll::Ready(GeneratorState::Yielded(x)) => Poll::Ready(Some(x)),
            Poll::Ready(GeneratorState::Complete(())) => Poll::Ready(None),
            Poll::Pending => Poll::Pending,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::{
        stack::{let_gen_using, Co},
        testing::SlowFuture,
    };
    use futures_executor::block_on_stream;
    use futures_util::stream;

    #[test]
    fn blocking() {
        async fn produce(mut co: Co<'_, i32>) {
            co.yield_(10).await;
            co.yield_(20).await;
        }

        let_gen_using!(gen, produce);
        let stream = stream::iter(gen);
        let items: Vec<_> = block_on_stream(stream).collect();
        assert_eq!(items, [10, 20]);
    }

    #[test]
    fn non_blocking() {
        async fn produce(mut co: Co<'_, i32>) {
            SlowFuture::new().await;
            co.yield_(10).await;
            SlowFuture::new().await;
            co.yield_(20).await;
        }

        let_gen_using!(gen, produce);
        let items: Vec<_> = block_on_stream(gen).collect();
        assert_eq!(items, [10, 20]);
    }
}
