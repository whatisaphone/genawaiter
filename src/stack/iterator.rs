use crate::{ops::GeneratorState, stack::generator::Gen};
use core::future::Future;

impl<'s, Y, F: Future<Output = ()>> IntoIterator for Gen<'s, Y, (), F> {
    type Item = Y;
    type IntoIter = IntoIter<'s, Y, F>;

    #[must_use]
    fn into_iter(self) -> Self::IntoIter {
        IntoIter { generator: self }
    }
}

pub struct IntoIter<'s, Y, F: Future<Output = ()>> {
    generator: Gen<'s, Y, (), F>,
}

impl<'s, Y, F: Future<Output = ()>> Iterator for IntoIter<'s, Y, F> {
    type Item = Y;

    fn next(&mut self) -> Option<Self::Item> {
        match self.generator.resume(()) {
            GeneratorState::Yielded(x) => Some(x),
            GeneratorState::Complete(()) => None,
        }
    }
}

impl<'r, 's, Y, F: Future<Output = ()>> IntoIterator for &'r mut Gen<'s, Y, (), F> {
    type Item = Y;
    type IntoIter = MutIntoIter<'r, 's, Y, F>;

    fn into_iter(self) -> Self::IntoIter {
        MutIntoIter { generator: self }
    }
}

pub struct MutIntoIter<'r, 's, Y, F: Future<Output = ()>> {
    generator: &'r mut Gen<'s, Y, (), F>,
}

impl<'r, 's, Y, F: Future<Output = ()>> Iterator for MutIntoIter<'r, 's, Y, F> {
    type Item = Y;

    fn next(&mut self) -> Option<Self::Item> {
        match self.generator.resume(()) {
            GeneratorState::Yielded(x) => Some(x),
            GeneratorState::Complete(()) => None,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::stack::{let_gen_using, Co, Gen, Shelf};
    use core::iter::IntoIterator;

    async fn produce(mut co: Co<'_, i32>) {
        co.yield_(10).await;
        co.yield_(20).await;
    }

    #[test]
    fn let_gen_using_into_iter() {
        let_gen_using!(gen, produce);

        let items: Vec<_> = gen.into_iter().collect();
        assert_eq!(items, [10, 20]);
    }

    #[test]
    fn let_gen_using_for_loop() {
        let_gen_using!(gen, produce);

        let mut sum = 0;
        for x in gen {
            sum += x;
        }
        assert_eq!(sum, 30);
    }

    #[test]
    fn shelf_generator_into_iter() {
        let mut shelf = Shelf::new();
        let gen = unsafe { Gen::new(&mut shelf, produce) };

        let items: Vec<_> = gen.into_iter().collect();
        assert_eq!(items, [10, 20]);
    }

    #[test]
    fn shelf_generator_for_loop() {
        let mut shelf = Shelf::new();
        let gen = unsafe { Gen::new(&mut shelf, produce) };

        let mut sum = 0;
        for x in gen {
            sum += x;
        }
        assert_eq!(sum, 30);
    }
}
