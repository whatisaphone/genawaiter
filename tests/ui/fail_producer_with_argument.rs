use genawaiter::stack::producer_fn;

#[producer_fn(u8)]
async fn odds(co: Co<'_, u8>) {
    co.yield_(10).await;
}

fn main() {}
