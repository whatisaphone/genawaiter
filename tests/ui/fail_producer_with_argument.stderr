error: A generator producer cannot accept any arguments. Instead, consider using a closure and capturing the values you need.
 --> $DIR/fail_producer_with_argument.rs:4:1
  |
4 | async fn odds(co: Co<'_, u8>) {
  | ^^^^^
